beacon-chain/monitor/service.go

```diff
@@ -59,6 +84,187 @@ type Service struct {
 	lastSyncedEpoch             types.Epoch
 }
 
+// NewService sets up a new validator monitor instance when given a list of
+// validator indices to track.
+func NewService(ctx context.Context, config *ValidatorMonitorConfig, tracked []types.ValidatorIndex) (*Service, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	r := &Service{
+		config:                      config,
+		ctx:                         ctx,
+		cancel:                      cancel,
+		TrackedValidators:           make(map[types.ValidatorIndex]interface{}, len(tracked)),
+		latestPerformance:           make(map[types.ValidatorIndex]ValidatorLatestPerformance),
+		aggregatedPerformance:       make(map[types.ValidatorIndex]ValidatorAggregatedPerformance),
+		trackedSyncCommitteeIndices: make(map[types.ValidatorIndex][]types.CommitteeIndex),
+	}
+	for _, idx := range tracked {
+		r.TrackedValidators[idx] = nil
+	}
+	return r, nil
+}
+
+// Start waits until the beacon is synced and starts the monitoring system.
+func (s *Service) Start() {
+	tracked := make([]types.ValidatorIndex, len(s.TrackedValidators))
+	i := 0
+	for idx := range s.TrackedValidators {
+		tracked[i] = idx
+		i++
+	}
+	sort.Slice(tracked, func(i, j int) bool { return tracked[i] < tracked[j] })
+	log.WithFields(logrus.Fields{
+		"ValidatorIndices": tracked,
+	}).Info("Started service")
+	s.isRunning = false
+	stateChannel := make(chan *feed.Event, 1)
+	stateSub := s.config.StateNotifier.StateFeed().Subscribe(stateChannel)
+
+	go func() {
+		if err := s.waitForSync(stateChannel, stateSub); err != nil {
+			log.WithError(err)
+			return
+		}
+		state, err := s.config.HeadFetcher.HeadState(s.ctx)
+		if err != nil {
+			log.WithError(err).Error("Could not get head state")
+			return
+		}
+		if state == nil {
+			log.Error("Head state is nil")
+			return
+		}
+		epoch := slots.ToEpoch(state.Slot())
+		log.WithField("Epoch", epoch).Debug("Synced to head epoch, starting reporting performance")
+
+		s.Lock()
+		for idx := range s.TrackedValidators {
+			balance, err := state.BalanceAtIndex(idx)
+			if err != nil {
+				log.WithError(err).WithField("ValidatorIndex", idx).Error(
+					"Could not fetch starting balance, skipping aggregated logs.")
+				balance = 0
+			}
+			s.aggregatedPerformance[idx] = ValidatorAggregatedPerformance{
+				startEpoch:   epoch,
+				startBalance: balance,
+			}
+			s.latestPerformance[idx] = ValidatorLatestPerformance{
+				balance: balance,
+			}
+		}
+		s.Unlock()
+		s.updateSyncCommitteeTrackedVals(state)
+		s.Lock()
+		s.isRunning = true
+		s.Unlock()
```

Comment:
I dont think these should be in the background routine.

Reply:
I couldn't solve the tests without sending these to the background, there are lots of races when subscribing to the channels.